### ISSUE TITLE
make error: "panic: assertion cv8_state.source_files != NULL"

### ISSUE THREAD

**Issue by Softwareminister** [openssl/openssl#11532]
When compiling openssl on Win64 with MS Visual Studio 2019, I receive reproducably the error

    ```
crypto\aes\aes-x86_64.asm: panic: assertion cv8_state.source_files != NULL failed at output/codeview.c:517
    ```

openssl version is 1.1.1 (current download from today)
NASM version is 2.14.02, also lastest available

The Configuration step of openssl seems to work well, output is as follows:
    ```
Z:\projects\Tools\openssl>perl Configure VC-WIN64A
Configuring OpenSSL version 3.0.0-dev for target VC-WIN64A
Using os-specific seed configuration
Creating configdata.pm
Running configdata.pm
Creating makefile

**********************************************************************
***                                                                ***
***   OpenSSL has been successfully configured                     ***
***                                                                ***
***   If you encounter a problem while building, please open an    ***
***   issue on GitHub <https://github.com/openssl/openssl/issues>  ***
***   and include the output from the following command:           ***
***                                                                ***
***       perl configdata.pm --dump                                ***
***                                                                ***
***   (If you are new to OpenSSL, you might want to consult the    ***
***   'Troubleshooting' section in the INSTALL file first)         ***
***                                                                ***
**********************************************************************
    ```

Full output on the buggy compilation is:
    ```
Z:\projects\Tools\openssl>nmake -f makefile

Microsoft (R) Program Maintenance Utility Version 14.24.28314.0
Copyright (C) Microsoft Corporation.  All rights reserved.

        "C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.24.28314\bin\HostX86\x86\nmake.exe" /                   depend && "C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.24.28314\bin\HostX86\x86\nmake.exe" /                   _build_sw

Microsoft (R) Program Maintenance Utility Version 14.24.28314.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Microsoft (R) Program Maintenance Utility Version 14.24.28314.0
Copyright (C) Microsoft Corporation.  All rights reserved.

        "nasm"  -Ox -f win64 -DNEAR -g -o crypto\aes\libcrypto-shlib-aes-x86_64.obj "crypto\aes\aes-x86_64.asm"
crypto\aes\aes-x86_64.asm: panic: assertion cv8_state.source_files != NULL failed at output/codeview.c:517
NMAKE : fatal error U1077: '"C:\Program Files\NASM\nasm.EXE"' : return code '0x3'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.24.28314\bin\HostX86\x86\nmake.exe"' : return code '0x2'
Stop.
    ```

The dump of the config is as follows:
    ```
Z:\projects\Tools\openssl>perl configdata.pm --dump

Command line (with current working directory = .):

    C:\Perl64\bin\perl.exe Configure VC-WIN64A

Perl information:

    C:\Perl64\bin\perl.exe
    5.28.1 for MSWin32-x64-multi-thread

Enabled features:

    aria
    asm
    async
    autoalginit
    autoerrinit
    autoload-config
    bf
    blake2
    camellia
    capieng
    cast
    chacha
    cmac
    cmp
    cms
    comp
    ct
    deprecated
    des
    dgram
    dh
    dsa
    dso
    dtls
    dynamic-engine
    ec
    ec2m
    ecdh
    ecdsa
    engine
    err
    filenames
    fips
    gost
    idea
    legacy
    makedepend
    md4
    mdc2
    module
    multiblock
    nextprotoneg
    pinshared
    ocb
    ocsp
    padlockeng
    pic
    poly1305
    posix-io
    psk
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    secure-memory
    seed
    shared
    siphash
    siv
    sm2
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    static-engine
    stdio
    tests
    threads
    tls
    ts
    ui-console
    uplink
    whirlpool
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method

Disabled features:

    afalgeng            [not-linux] OPENSSL_NO_AFALGENG
    asan                [default]   OPENSSL_NO_ASAN
    buildtest-c++       [default]
    crypto-mdebug       [default]   OPENSSL_NO_CRYPTO_MDEBUG
    devcryptoeng        [default]   OPENSSL_NO_DEVCRYPTOENG
    ec_nistp_64_gcc_128 [default]   OPENSSL_NO_EC_NISTP_64_GCC_128
    egd                 [default]   OPENSSL_NO_EGD
    external-tests      [default]   OPENSSL_NO_EXTERNAL_TESTS
    fuzz-libfuzzer      [default]   OPENSSL_NO_FUZZ_LIBFUZZER
    fuzz-afl            [default]   OPENSSL_NO_FUZZ_AFL
    ktls                [default]   OPENSSL_NO_KTLS
    md2                 [default]   OPENSSL_NO_MD2 (skip crypto\md2)
    msan                [default]   OPENSSL_NO_MSAN
    rc5                 [default]   OPENSSL_NO_RC5 (skip crypto\rc5)
    sctp                [default]   OPENSSL_NO_SCTP
    ssl-trace           [default]   OPENSSL_NO_SSL_TRACE
    trace               [default]   OPENSSL_NO_TRACE
    ubsan               [default]   OPENSSL_NO_UBSAN
    unit-test           [default]   OPENSSL_NO_UNIT_TEST
    weak-ssl-ciphers    [default]   OPENSSL_NO_WEAK_SSL_CIPHERS
    zlib                [default]
    zlib-dynamic        [default]
    ssl3                [default]   OPENSSL_NO_SSL3
    ssl3-method         [default]   OPENSSL_NO_SSL3_METHOD

Config target attributes:

    AR => "lib",
    ARFLAGS => "/nologo",
    AS => "nasm",
    ASFLAGS => "-g",
    CC => "cl",
    CFLAGS => "/W3 /wd4090 /nologo /O2",
    CPP => "\$(CC) /EP /C",
    HASHBANGPERL => "/usr/bin/env perl",
    LD => "link",
    LDFLAGS => "/nologo /debug",
    MT => "mt",
    MTFLAGS => "-nologo",
    RANLIB => "CODE(0xe3d1a0)",
    RC => "rc",
    aroutflag => "/out:",
    asflags => "-Ox -f win64 -DNEAR",
    asm_arch => "x86_64",
    asoutflag => "-o ",
    bin_cflags => "/Zi /Fdapp.pdb",
    bin_lflags => "/subsystem:console /opt:ref",
    bn_ops => "SIXTY_FOUR_BIT",
    build_file => "makefile",
    build_scheme => [ "unified", "windows", "VC-common" ],
    cflags => "/Gs0 /GF /Gy /MD",
    coutflag => "/Fo",
    cppflags => "",
    defines => [ "OPENSSL_SYS_WIN32", "WIN32_LEAN_AND_MEAN", "UNICODE", "_UNICODE", "_CRT_SECURE_NO_DEPRECATE", "_WINSOCK_DEPRECATED_NO_WARNINGS" ],
    disable => [  ],
    dso_cflags => "/Zi /Fddso.pdb",
    dso_scheme => "win32",
    enable => [  ],
    ex_libs => "ws2_32.lib gdi32.lib advapi32.lib crypt32.lib user32.lib",
    includes => [  ],
    ldoutflag => "/out:",
    lflags => "",
    lib_cflags => "/Zi /Fdossl_static.pdb",
    lib_cppflags => "",
    lib_defines => [ "L_ENDIAN" ],
    module_cflags => "",
    module_cxxflags => undef,
    module_ldflags => "/dll",
    mtinflag => "-manifest ",
    mtoutflag => "-outputresource:",
    multilib => "-x64",
    perl_platform => "Windows::MSVC",
    perlasm_scheme => "auto",
    rcoutflag => "/fo",
    shared_cflag => "",
    shared_defflag => "",
    shared_defines => [  ],
    shared_ldflag => "/dll",
    shared_rcflag => "",
    shared_target => "win-shared",
    sys_id => "WIN64A",
    thread_defines => [  ],
    thread_scheme => "winthreads",
    unistd => "<unistd.h>",
    uplink_arch => "x86_64",

Recorded environment:

    AR =
    ARFLAGS =
    AS =
    ASFLAGS =
    BUILDFILE =
    CC =
    CFLAGS =
    CPP =
    CPPDEFINES =
    CPPFLAGS =
    CPPINCLUDES =
    CROSS_COMPILE =
    CXX =
    CXXFLAGS =
    HASHBANGPERL =
    LD =
    LDFLAGS =
    LDLIBS =
    MT =
    MTFLAGS =
    OPENSSL_LOCAL_CONFIG_DIR =
    PERL =
    RANLIB =
    RC =
    RCFLAGS =
    RM =
    WINDRES =
    __CNF_CFLAGS =
    __CNF_CPPDEFINES =
    __CNF_CPPFLAGS =
    __CNF_CPPINCLUDES =
    __CNF_CXXFLAGS =
    __CNF_LDFLAGS =
    __CNF_LDLIBS =

Makevars:

    AR              = lib
    ARFLAGS         = /nologo
    AS              = nasm
    ASFLAGS         = -g
    CC              = cl
    CFLAGS          = /W3 /wd4090 /nologo /O2
    CPP             = $(CC) /EP /C
    CPPDEFINES      =
    CPPFLAGS        =
    CPPINCLUDES     =
    CXXFLAGS        =
    HASHBANGPERL    = /usr/bin/env perl
    LD              = link
    LDFLAGS         = /nologo /debug
    LDLIBS          =
    MT              = mt
    MTFLAGS         = -nologo
    PERL            = C:\Perl64\bin\perl.exe
    RC              = rc
    RCFLAGS         =

NOTE: These variables only represent the configuration view.  The build file
template may have processed these variables further, please have a look at the
build file for more exact data:
    makefile

build file:

    makefile

build file templates:

    Configurations\common0.tmpl
    Configurations\windows-makefile.tmpl
    Configurations\common.tmpl

    ```

Any help appreciated.

**Comment by mattcaswell** [comment]
This looks like an internal assertion failure inside NASM, i.e. a bug in NASM (output/codeview.c: is not an OpenSSL file - its a NASM one: https://fossies.org/linux/nasm/output/codeview.c). You might want to raise an issue there.

**Comment by Softwareminister** [comment]
Thanks mattcaswell, I will do so. Sadly, openssl cannot be put into work without NASM on windows.

**Comment by mattcaswell** [comment]
As a workaround perhaps try downgrading NASM to some older version. Perhaps not all versions are affected?

**Comment by Softwareminister** [comment]
Tried the latest realease candidate, and also version 2.12 from early 2016, all behave the same (errornoeous)

**Comment by emmenlau** [comment]
I'm under the impression that this issue is the same as https://bugzilla.nasm.us/show_bug.cgi?id=3392658

I can confirm that this is fixed with nasm 2.16.01 (current latest) when building OpenSSL 1.1.1s (current latest). Possibly older versions also work, but I did not test that.

**Comment by emmenlau** [comment]
I think this issue can be closed as fixed upstream, and since it seems to be a nasm issue, it is anyways out of scope for OpenSSL...

**Comment by t8m** [comment]
Thank you. Closing.

**Comment by snellejelle99** [comment]
> I'm under the impression that this issue is the same as https://bugzilla.nasm.us/show_bug.cgi?id=3392658
> 
> I can confirm that this is fixed with nasm 2.16.01 (current latest) when building OpenSSL 1.1.1s (current latest). Possibly older versions also work, but I did not test that.

Im currently trying to compile openssl 3.07 with NASM 2.16.01 and am receiving a similar error (though slightly different)
```
"nasm"  -Ox -f win64 -DNEAR -g -o crypto\aes\libcrypto-shlib-aes-x86_64.obj "crypto\aes\aes-x86_64.asm"
panic: crypto\aes\aes-x86_64.asm: assertion cv8_state.source_files != NULL failed at output/codeview.c:515
NMAKE : fatal error U1077: '"C:\Program Files\NASM\nasm.EXE"' : return code '0x3'
```

**Comment by dgm3333** [comment]
I'm getting the same issue. I've reported to NASM devs too.

Using OpenSSL master (ie latest) and NASM 2.16.01 (ie also latest) on Win10 x64 with an AMD Ryzen 9 processor

I'm trying to compile OpenSSL on Win64 but it fails with the following error:
panic: crypto\aes\aes-x86_64.asm: assertion cv8_state.source_files != NULL failed at output/codeview.c:515

Seems to be same issue as 
https://lists.nasm.us/archives/nasm-bugs/2021-February/002048.html
and
https://github.com/openssl/openssl/issues/11532
  in this issue SSL devs state this is an "upstream" (NASM) problem.

The following is the output from the Visual Studio prompt in case it's of any use.

        cmd /C ""cl" /Zs /showIncludes  /Zi /Fdossl_static.pdb /Gs0 /GF /Gy /MD /W3 /wd4090 /nologo /O2 -I"." -I"include" -I"apps\include" -D"L_ENDIAN" -D"OPENSSL_PIC" -D"OPENSSLDIR=\"C:\\Program Files\\Common Files\\SSL\"" -D"ENGINESDIR=\"C:\\Program Files\\OpenSSL\\lib\\engines-3\"" -D"MODULESDIR=\"C:\\Program Files\\OpenSSL\\lib\\ossl-modules\"" -D"OPENSSL_BUILDING_OPENSSL" -D"OPENSSL_SYS_WIN32" -D"WIN32_LEAN_AND_MEAN" -D"UNICODE" -D"_UNICODE" -D"_CRT_SECURE_NO_DEPRECATE" -D"_WINSOCK_DEPRECATED_NO_WARNINGS" -D"NDEBUG"   "apps\lib\tlssrp_depr.c" > apps\lib\libapps-lib-tlssrp_depr.d 2>&1"
        "lib" /nologo /out:apps\libapps.lib @C:\Users\dgm55\AppData\Local\Temp\nm2AED.tmp
        "nasm"  -Ox -f win64 -DNEAR -g -o crypto\aes\libcrypto-shlib-aes-x86_64.obj "crypto\aes\aes-x86_64.asm"
panic: crypto\aes\aes-x86_64.asm: assertion cv8_state.source_files != NULL failed at output/codeview.c:515
NMAKE : fatal error U1077: '"C:\Program Files\NASM\nasm.EXE"' : return code '0x3'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\bin\HostX86\x86\nmake.exe"' : return code '0x2'
Stop.
